### ISSUE TITLE
Cutout2D partial behavior with mask

### ISSUE THREAD

**Issue by wtgee** [astropy/astropy#7659]
Hello, I'm not sure if this is expected behavior but it threw me for a bit. `astropy.nddata.Cutout2D` works fine for data that is masked unless you are using `mode="partial"`, in which cases it seems to ignore the mask. Trivial example:

```
>>> import numpy as np
>>> from astropy.nddata import Cutout2D
>>> foo = np.random.randint(0, 1000, 36).reshape(6, 6)
[[362 291 495 381 915 328]
 [809 476   3  67 368 598]
 [910 936 782 677  31 988]
 [ 69 541 941 232 585 354]
 [431 219 140 673  75 224]
 [136 648 530 754 228 391]]

>>> bar = np.ma.array(foo, mask=foo > 300)
[[-- 291 -- -- -- --]
 [-- -- 3 67 -- --]
 [-- -- -- -- 31 --]
 [69 -- -- 232 -- --]
 [-- 219 140 -- 75 224]
 [136 -- -- -- 228 --]]

>>> Cutout2D(bar, (3, 3), size=5, mode='partial', fill_value=0).data
[[-- 3 67 -- --]
 [-- -- -- 31 --]
 [-- -- 232 -- --]
 [219 140 -- 75 224]
 [-- -- -- 228 --]]

>>> Cutout2D(bar, (4, 3), size=5, mode='partial', fill_value=0).data
[[  3  67 368 598   0]
 [782 677  31 988   0]
 [941 232 585 354   0]
 [140 673  75 224   0]
 [530 754 228 391   0]]
```
It's the last case that seems odd to me. If I've masked a bad pixel in there I would expect that to stay masked even with the `fill_value`. 

Is this expected behavior?

**Comment by pllim** [comment]
@mwcraig or @MSeifert04 ?

**Comment by MSeifert04** [comment]
Good question. I don't think we actually support masked arrays there. Maybe accidentally but definitely not on purpose (there are no test cases for masked data and no docs on masked data).

I wouldn't mind supporting masked arrays (if we can support them relatively easily) if someone wants to step up and completely checks the masked-array compatibility for `Cutout2D` and fix the "problems". But I don't know how much work that'll be and I'm a bit stretched currently (no time to do this myself).

**Comment by wtgee** [comment]
Ok, thanks @MSeifert04. I suppose for now the best plan would be to do the cutout and then use one of the `bbox_original` properties to apply the mask after the fact. 

I'll read through the docs and see if there isn't a good place to add some kind of sentence or note that mentions that masks are not explicitly supported yet. And maybe if I magically find time I'll take a stab at looking over the related code.

I will leave open in case it's helpful but feel free to close.  Cheers!

**Comment by MSeifert04** [comment]
@wtgee I don't think we should close this. I think there are use-cases for masked array cutouts.

It's just that it's not a bug. However it's a totally fine feature-request. :)
